testsuite/pytests/sli2py_mpi/mpi_test_wrapper.py

```diff
@@ -0,0 +1,197 @@
+    MULTI_LABEL = "multi-{}"
+    OTHER_LABEL = "other-{}"
+
+    RUNNER_TEMPLATE = textwrap.dedent(
+        """\
+        SPIKE_LABEL = '{spike_lbl}'
+        MULTI_LABEL = '{multi_lbl}'
+        OTHER_LABEL = '{other_lbl}'
+
+        {fcode}
+
+        if __name__ == '__main__':
+            {fname}({params})
+        """
+    )
+
+    def __init__(self, procs_lst, debug=False):
+        try:
+            iter(procs_lst)
+        except TypeError:
+            raise TypeError("procs_lst must be a list of numbers")
+
+        self._procs_lst = procs_lst
+        self._debug = debug
+        self._spike = None
+        self._multi = None
+        self._other = None
+
+    def _func_without_decorators(self, func):
+        return "".join(line for line in inspect.getsourcelines(func)[0] if not line.startswith("@"))
+
+    def _params_as_str(self, *args, **kwargs):
+        return ", ".join(
+            part
+            for part in (
+                ", ".join(f"{arg}" for arg in args),
+                ", ".join(f"{key}={value}" for key, value in kwargs.items()),
+            )
+            if part
+        )
+
+    def _write_runner(self, tmpdirpath, func, *args, **kwargs):
+        with open(tmpdirpath / self.RUNNER, "w") as fp:
+            fp.write(
+                self.RUNNER_TEMPLATE.format(
+                    spike_lbl=self.SPIKE_LABEL,
+                    multi_lbl=self.MULTI_LABEL,
+                    other_lbl=self.OTHER_LABEL,
+                    fcode=self._func_without_decorators(func),
+                    fname=func.__name__,
+                    params=self._params_as_str(*args, **kwargs),
+                )
+            )
+
+    def __call__(self, func):
+        def wrapper(func, *args, **kwargs):
+            # "delete" parameter only available in Python 3.12 and later
+            try:
+                tmpdir = tempfile.TemporaryDirectory(delete=not self._debug)
+            except TypeError:
+                tmpdir = tempfile.TemporaryDirectory()
+
+            # TemporaryDirectory() is not os.PathLike, so we need to define a Path explicitly
+            # To ensure that tmpdirpath has the same lifetime as tmpdir, we define it as a local
+            # variable in the wrapper() instead of as an attribute of the decorator.
+            tmpdirpath = Path(tmpdir.name)
+            self._write_runner(tmpdirpath, func, *args, **kwargs)
+
+            res = {}
+            for procs in self._procs_lst:
+                res[procs] = subprocess.run(
+                    ["mpirun", "-np", str(procs), "--oversubscribe", "python", self.RUNNER],
+                    check=True,
+                    cwd=tmpdirpath,
+                    capture_output=self._debug,
+                )
+
+            if self._debug:
+                print("\n\n")
+                print(res)
+                print(f"\n\nTMPDIR: {tmpdirpath}\n\n")
+
+            self.assert_correct_results(tmpdirpath)
+
+        return decorator(wrapper, func)
```

Comment:
This seems to work, the only difference seems to be that the call signature of the wrapped function is different from the original, while `decorator` preserves it (see https://github.com/micheles/decorator/blob/master/docs/documentation.md#statement-of-the-problem). But since this does not cause a problem for us, I agree that avoiding a new dependency is the greater advantage.